doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,917 @@
+   via a LaunchDarkly flag.
+2. Incremental progress. If we write both formats then we could read back Arrow
+   data but sort and consolidate with the `Codec` format. Currently decoding
+   protobuf (i.e. `ProtoDatum`) is computationally expensive, getting an
+   incremental win of improving decode time would be nice.
+3. Doesn't increase the number of `PUT` or `GET` requests to `S3`.
+4. Doesn't require any change to state stored in Consensus.
+5. (Forward looking) When we switch to reading only Arrow data we can leverage
+   Parquet's projection pushdown to ignore the old `'k'` and `'v'` columns.
+
+The current max blob size is 128MiB, we have two options for how we want to
+handle the "max" size when we write both `Codec` data and Arrow at the same
+time.
+
+1. Ignore Arrow when calculating the max. This increases the size of S3 objects
+   which may impact performance for things that have been tuned to assume
+   128MiB sized objets, e.g. compute back pressure.
+2. Include Arrow when calculating the max. This reduces the amount of rows that
+   can get stored in a single blob, thus increasing the number of Parts
+   per-Run. This increases the number of S3 `GET`s and `PUT`s which could
+   impact cost, and requires `ConsolidatingIter` to handle a larger number of
+   Parts.
+
+Neither option is great, but I am leanning towards [1] since it puts the
+pressure/complexity on S3 and the network more than Persist internals.
+
+### Consolidation and Compaction
+
+To start we can write both the `Codec` and Arrow versions of the data, and do
+compaction based on the bytes from `Codec`. Eventually we'll need to
+consolidate the Arrow data directly which requires sorting the data. Sorting
+an Arrow `StructArray` is possible (e.g. the `pyarrow` library [supports it](https://arrow.apache.org/docs/python/generated/pyarrow.StructArray.html#pyarrow.StructArray.sort))
+but neither the `arrow-rs` or `arrow2` crates support it. I filed an
+[issue with `arrow-rs`](https://github.com/apache/arrow-rs/issues/5559) but
+realistically this is something we'll probably need to implement. Once data is
+sorted consolidating an Arrow array is relatively easy and fast! To delete a
+value all we need to do is unset the bit in the
+[validity bitmap](https://arrow.apache.org/docs/format/Columnar.html#validity-bitmaps)
+for that value.
+
+#### Across Multiple Versions of Data
+
+There are three different scenarios we need to handle consolidation for:
+
+1. Evolving the way we encode `Datum`s to Arrow Arrays.
+2. Batch A with `Codec` data, Batch B with Arrow Arrays.
+3. Two batches across different versions of a table (e.g. after `ALTER TABLE .. ADD COLUMN ...`)
+
+For all of these scenarios I believe we can handle consolidation by first
+decoding to the Rust type (e.g. `Row`) and consolidating based on that. It will
+be relatively inefficient, but none of these scenarios should happen too often.
```

Comment:
This will leave data in an order that doesn't match either the existing consolidated output (based on bytes) or the eventual-future output (based on the structured representation).

I'd pitch just sorting based the `Codec` representation while we have it! The two notes below seem to handle things well.

Reply:
> I'd pitch just sorting based the Codec representation while we have it! The two notes below seem to handle things well.

Totally agree! These are scenarios in which we don't have `Codec` data, or we're trying to consolidate `Codec` with `Arrow`. Although thinking about this and streaming consolidation, made me realize the best approach might be decoding to the Rust type `K` and then encoding into the latest Arrow format. Will need to think on this some more and get back to you.